### ISSUE TITLE
Reproducible results from Optuna multi-layer perceptron -pytorch

### ISSUE THREAD

**Issue by pchunduru10** [optuna/optuna#3537]
### Environment

- Optuna version:
- Python version: 3.6
- OS: Ubuntu
- (Optional) Other libraries and their versions:

### Description

I am currently using Optuna (TPSE Sampler) for the multi-layer perceptron regression model. I tuning on parameters including , learning rate, num hidden layers, num of nodes in hidden layer, dropout, optimizer (Adam, RMSprop, SGD) .

The objective function I am using is MSE for test samples

I am trying to see of the best parameters reported from optuna for n_trails are consistent if we increase the trial or re-run for same number of trials.
I see that the hidden layer size changes randomly for the same n_trails run repeatedly. or even increasing the n_trials.
Ex n_trials range  = 100, 500 , 1000, 1500, 2000

### Error messages, stack traces, or logs

```shell
Inconsistent hidden layer size and hidden layer nodes for same number of optuna trials.
```

### Steps to reproduce

1. Fixed random seed ( np.random.seed(seed),  torch.manual_seed(SEED) )
2. Fixed  train data loader to run on entire data instead of mini-batch training (I have a small sample size- so no computation issue on training entire training data)

```python
# python code
```

**Comment by nzw0301** [comment]
Hi, could you share minimal reproducible codes with us?

**Comment by pchunduru10** [comment]
```
"""import libraries."""
import os, sys
import argparse

import numpy as np
import pandas as pd

import json
import optuna
import torch
import torch.optim as optim
from torch.utils.data import DataLoader
from sklearn.datasets import load_boston
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection  import train_test_split

torch.manual_seed(42)


class BostonDataset(torch.utils.data.Dataset):
  '''
  Prepare the Boston dataset for regression
  '''

  def __init__(self, X, y, scale_data=True):
    if not torch.is_tensor(X) and not torch.is_tensor(y):
      # Apply scaling if necessary
      if scale_data:
          X = StandardScaler().fit_transform(X)
      self.X = torch.from_numpy(X)
      self.y = torch.from_numpy(y)

  def __len__(self):
      return len(self.X)

  def __getitem__(self, i):
      return self.X[i], self.y[i]



def MLP(trial, in_features, n_layers, dropout, n_output):
    """
    Generate sequential network with OPTUNA bayesian optimization.


    :param trial: optuna trial class
    :param in_features: num of input nodes
    :param n_layers: num of hidden layers
    :param dropout: perc of final layer dropout
    :param n_output: num of output nodes
    :return: sequential multi layer perceptron model 
    """
    layers = []
    fc_layer = in_features
    
    for i in range(n_layers):
        
        out_features = trial.suggest_int("n_units_l{}".format(i), 2, in_features)
        layers.append(torch.nn.Linear(in_features, out_features))
        layers.append(torch.nn.LeakyReLU())
        in_features = out_features

    layers.append(torch.nn.Linear(in_features, fc_layer)) 
    layers.append(torch.nn.LeakyReLU())

    layers.append(torch.nn.Dropout(dropout))
    layers.append(torch.nn.Linear(fc_layer,n_output)) 
    
    return torch.nn.Sequential(*layers)


def train_net(trial, params):
    # Load Boston dataset
    X, y = load_boston(return_X_y=True)
    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)

    # Prepare Boston dataset
    trainloader = torch.utils.data.DataLoader(BostonDataset(X_train, y_train), batch_size=10, shuffle=True, num_workers=1)
    testloader = torch.utils.data.DataLoader(BostonDataset(X_test, y_test), batch_size=1, shuffle=True, num_workers=1)
    
    # Initialize the MLP
    net = MLP(trial,
            in_features= X.shape[1],
            n_layers=params['n_layers'] ,
            dropout=params['dropout'],
            n_output= 1)
    
  # Define the loss function and optimizer
    loss_function = torch.nn.MSELoss()
    optimizer = getattr(optim, params['optimizer'])(net.parameters(), lr= params['learning_rate'], weight_decay=params['weight_decay'])
  
    # Run the training loop
    for epoch in range(0, 5): # 5 epochs at maximum
        
        total_test_loss = []
        # Print epoch
        print(f'Starting epoch {epoch+1}')
        
        # Set current loss value
        current_loss = 0.0
        
        # Iterate over the DataLoader for training data
        for i, data in enumerate(trainloader, 0):
        
            # Get and prepare inputs
            inputs, targets = data
            inputs, targets = inputs.float(), targets.float()
            targets = targets.reshape((targets.shape[0], 1))
            
            # Zero the gradients
            optimizer.zero_grad()
            # Perform forward pass
            outputs = net(inputs)
            # Compute loss
            loss = loss_function(outputs, targets)
            # Perform backward pass
            loss.backward()
            # Perform optimization
            optimizer.step()
            
        net.eval()
        with torch.no_grad():
            for i, data in enumerate(testloader, 0):
                # Get and prepare inputs
                inputs, targets = data
                inputs, targets = inputs.float(), targets.float()
                targets = targets.reshape((targets.shape[0], 1))
                
                # # Perform forward pass
                test_outputs = net(inputs)
                test_loss = loss_function(test_outputs, targets)
                total_test_loss.append(test_loss.item())
        

    # Process is complete.
    return total_test_loss


def objective(trial):
    """
    Objective function to run bayesian hyperparameter tuning.

    :param trial: optuna study
    :param checkpoint_dir: checkpoint dir args
    :param cfg: config file
    :return: mean RMSE test loss
    """ 
    # For TPESampler 
    params = {
              'learning_rate': trial.suggest_loguniform('learning_rate', 1e-6, 1e-2), 
              'optimizer': trial.suggest_categorical("optimizer", ["Adam", "RMSprop", "SGD"]),
              'weight_decay': trial.suggest_loguniform('weight_decay', 1e-4, 1e-2),
              "n_layers" : trial.suggest_int("n_layers", 1, 4),
              "dropout" : trial.suggest_float('dropout',0.1,0.5,step = 0.1)
              }
  
    test_loss   = train_net(trial, params) 
    return np.mean(test_loss)# mean MSE loss


if __name__ == "__main__":

    parser = argparse.ArgumentParser()
    parser.add_argument('-o', '--optuna_trials', action='store', type= int,  default= 10, help="number of optuan trials")
    
    args = parser.parse_args()

    
    study = optuna.create_study(direction ="minimize", sampler=optuna.samplers.TPESampler())
    study.optimize(lambda trial: objective(trial), n_trials= args.optuna_trials)

    model_params = study.best_trial.params
    model_params['optuna_best_trial_number'] =  study.best_trial.number 
    model_params['optuna_best_trial_value'] = float( np.round(study.best_value,6) )
    model_params["n_trials"] = args.optuna_trials

    with open(f"data_{args.optuna_trials}.json", 'w') as fp:
        json.dump(model_params, fp)

```

**Comment by pchunduru10** [comment]
> ```
> """import libraries."""
> import os, sys
> import argparse
> 
> import numpy as np
> import pandas as pd
> 
> import json
> import optuna
> import torch
> import torch.optim as optim
> from torch.utils.data import DataLoader
> from sklearn.datasets import load_boston
> from sklearn.preprocessing import StandardScaler
> from sklearn.model_selection  import train_test_split
> 
> torch.manual_seed(42)
> 
> 
> class BostonDataset(torch.utils.data.Dataset):
>   '''
>   Prepare the Boston dataset for regression
>   '''
> 
>   def __init__(self, X, y, scale_data=True):
>     if not torch.is_tensor(X) and not torch.is_tensor(y):
>       # Apply scaling if necessary
>       if scale_data:
>           X = StandardScaler().fit_transform(X)
>       self.X = torch.from_numpy(X)
>       self.y = torch.from_numpy(y)
> 
>   def __len__(self):
>       return len(self.X)
> 
>   def __getitem__(self, i):
>       return self.X[i], self.y[i]
> 
> 
> 
> def MLP(trial, in_features, n_layers, dropout, n_output):
>     """
>     Generate sequential network with OPTUNA bayesian optimization.
> 
> 
>     :param trial: optuna trial class
>     :param in_features: num of input nodes
>     :param n_layers: num of hidden layers
>     :param dropout: perc of final layer dropout
>     :param n_output: num of output nodes
>     :return: sequential multi layer perceptron model 
>     """
>     layers = []
>     fc_layer = in_features
>     
>     for i in range(n_layers):
>         
>         out_features = trial.suggest_int("n_units_l{}".format(i), 2, in_features)
>         layers.append(torch.nn.Linear(in_features, out_features))
>         layers.append(torch.nn.LeakyReLU())
>         in_features = out_features
> 
>     layers.append(torch.nn.Linear(in_features, fc_layer)) 
>     layers.append(torch.nn.LeakyReLU())
> 
>     layers.append(torch.nn.Dropout(dropout))
>     layers.append(torch.nn.Linear(fc_layer,n_output)) 
>     
>     return torch.nn.Sequential(*layers)
> 
> 
> def train_net(trial, params):
>     # Load Boston dataset
>     X, y = load_boston(return_X_y=True)
>     X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=0)
> 
>     # Prepare Boston dataset
>     trainloader = torch.utils.data.DataLoader(BostonDataset(X_train, y_train), batch_size=10, shuffle=True, num_workers=1)
>     testloader = torch.utils.data.DataLoader(BostonDataset(X_test, y_test), batch_size=1, shuffle=True, num_workers=1)
>     
>     # Initialize the MLP
>     net = MLP(trial,
>             in_features= X.shape[1],
>             n_layers=params['n_layers'] ,
>             dropout=params['dropout'],
>             n_output= 1)
>     
>   # Define the loss function and optimizer
>     loss_function = torch.nn.MSELoss()
>     optimizer = getattr(optim, params['optimizer'])(net.parameters(), lr= params['learning_rate'], weight_decay=params['weight_decay'])
>   
>     # Run the training loop
>     for epoch in range(0, 5): # 5 epochs at maximum
>         
>         total_test_loss = []
>         # Print epoch
>         print(f'Starting epoch {epoch+1}')
>         
>         # Set current loss value
>         current_loss = 0.0
>         
>         # Iterate over the DataLoader for training data
>         for i, data in enumerate(trainloader, 0):
>         
>             # Get and prepare inputs
>             inputs, targets = data
>             inputs, targets = inputs.float(), targets.float()
>             targets = targets.reshape((targets.shape[0], 1))
>             
>             # Zero the gradients
>             optimizer.zero_grad()
>             # Perform forward pass
>             outputs = net(inputs)
>             # Compute loss
>             loss = loss_function(outputs, targets)
>             # Perform backward pass
>             loss.backward()
>             # Perform optimization
>             optimizer.step()
>             
>         net.eval()
>         with torch.no_grad():
>             for i, data in enumerate(testloader, 0):
>                 # Get and prepare inputs
>                 inputs, targets = data
>                 inputs, targets = inputs.float(), targets.float()
>                 targets = targets.reshape((targets.shape[0], 1))
>                 
>                 # # Perform forward pass
>                 test_outputs = net(inputs)
>                 test_loss = loss_function(test_outputs, targets)
>                 total_test_loss.append(test_loss.item())
>         
> 
>     # Process is complete.
>     return total_test_loss
> 
> 
> def objective(trial):
>     """
>     Objective function to run bayesian hyperparameter tuning.
> 
>     :param trial: optuna study
>     :param checkpoint_dir: checkpoint dir args
>     :param cfg: config file
>     :return: mean RMSE test loss
>     """ 
>     # For TPESampler 
>     params = {
>               'learning_rate': trial.suggest_loguniform('learning_rate', 1e-6, 1e-2), 
>               'optimizer': trial.suggest_categorical("optimizer", ["Adam", "RMSprop", "SGD"]),
>               'weight_decay': trial.suggest_loguniform('weight_decay', 1e-4, 1e-2),
>               "n_layers" : trial.suggest_int("n_layers", 1, 4),
>               "dropout" : trial.suggest_float('dropout',0.1,0.5,step = 0.1)
>               }
>   
>     test_loss   = train_net(trial, params) 
>     return np.mean(test_loss)# mean MSE loss
> 
> 
> if __name__ == "__main__":
> 
>     parser = argparse.ArgumentParser()
>     parser.add_argument('-o', '--optuna_trials', action='store', type= int,  default= 10, help="number of optuan trials")
>     
>     args = parser.parse_args()
> 
>     
>     study = optuna.create_study(direction ="minimize", sampler=optuna.samplers.TPESampler())
>     study.optimize(lambda trial: objective(trial), n_trials= args.optuna_trials)
> 
>     model_params = study.best_trial.params
>     model_params['optuna_best_trial_number'] =  study.best_trial.number 
>     model_params['optuna_best_trial_value'] = float( np.round(study.best_value,6) )
>     model_params["n_trials"] = args.optuna_trials
> 
>     with open(f"data_{args.optuna_trials}.json", 'w') as fp:
>         json.dump(model_params, fp)
> ```

Tried to create a unit test using a dummy BostonDataset. My actual script uses the same network and parameters that I am tuning on

**Comment by nzw0301** [comment]
Thanks. Could you try the code by following [this FAQ](https://optuna.readthedocs.io/en/stable/faq.html#how-can-i-obtain-reproducible-optimization-results), specifically specifying the random seed of TPE sampler.

**Comment by pchunduru10** [comment]
Thanks for quick response.Yes setting TPE seed reproduces the results. However if I increase the n_trials, I get the same results. So I am wondering if its actually find the best parameters fit.

**Comment by pchunduru10** [comment]
> 

n_trails = [500, 1000, 1500, 2000, 2500] for my actual script. It return one set of hyperparameters specifically hidden layers for 1000 trials, one for 1500 trials and so on. Althought the object function best value doesn't change much.

**Comment by nzw0301** [comment]
Thanks. The `n_trials` looks too large for hyperparameter search space, so Optuna might find an optimal value.

**Comment by pchunduru10** [comment]
> Thanks. The `n_trials` looks so large for hyperparameter search space, so Optuna might find an optimal value.

My thought is there is a drastic change in **hidden_layer** size across these trials. I tried running n_trails =1000, for 5 times with different TPE seed each time and it returns different hidden_layers. 
Same for n_trials = [1500, 2000, ] 

What could be the cause of the randomness ?

**Comment by nzw0301** [comment]
If the best objective values are similar among different seeds, the optimal hyperparameters are not unique; the reported results make sense for me.

**Comment by nzw0301** [comment]
Let me close this issue because I believe I've answered the question. But please feel free to report this issue.
